awx/main/tasks.py

```diff
@@ -396,6 +398,27 @@ def purge_old_stdout_files():
             logger.debug("Removing {}".format(os.path.join(settings.JOBOUTPUT_ROOT, f)))
 
 
+@task(queue=get_local_queuename)
+def cleanup_execution_environment_images():
+    if settings.IS_K8S:
+        return
+    images_in_use = [ee.image for ee in ExecutionEnvironment.objects.all()]
+    images_system = subprocess.run("podman images -a --format json".split(" "), capture_output=True)
+    if len(images_system.stdout) > 0:
+        images_system = json.loads(images_system.stdout)
+        for e in images_system:
+            if 'Names' in e:
+                image_name = e['Names'][0]
+            else:
+                image_name = e["Id"]
+            image_size = e['Size'] / 1e6
+            if image_name not in images_in_use:
```

Comment:
What if my EE has set `quay.io/project-receptor/receptor` but the reported output from `podman images -a --format json` gives:

```json
    {
        "Id": "bf4de35112248c7c5b1db46c43f87346f31e2717850831433af1fee6dfd0c2b4",
        "ParentId": "",
        "Size": 387129820,
        "Labels": {
            "license": "ASL2",
            "name": "receptor",
            "vendor": "project-receptor",
            "version": "0.9.1"
        },
        "Names": [
            "quay.io/project-receptor/receptor:latest"
        ],
        "Digest": "sha256:11d9a25f07ea27146d94772eb3c6f2faf5d5c75d9c3313fcad7390c7e417c32a",
        "Digests": [
            "sha256:11d9a25f07ea27146d94772eb3c6f2faf5d5c75d9c3313fcad7390c7e417c32a"
        ],
        "Created": 1602187764,
        "CreatedAt": "2020-10-08T20:09:24Z"
    }
```

I think it's valid to not include `:latest` so I'm wondering if it might confuse this code.

It's also going to be tricky to validate this. If it incorrectly deletes images and layers that it shouldn't, it will be quite difficult for QE to track down what was _successfully_ deleted, but shouldn't have been (triggering an expensive download on the next run).

Reply:
Good observation. We could assume that image names without a tag are using `latest` and check for those images on the system. The trick then is to find a reliable regex or something similar to parse the image name.

We could also add an image_id field to the EE model that tracks the ID of the images pulled, rather than image name, but tracking that id might be difficult.

Reply:
https://docs.docker.com/engine/reference/commandline/tag/

> A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits, underscores, periods and dashes. A tag name may not start with a period or a dash and may contain a maximum of 128 characters.

You'd need to rely on what you know is valid for the image URL & tag. I think it could work in exactly the way you say. It would be good to do some manual testing of different ways of specifying an image and making sure that this code still works as intended.

Reply:
to complicate matters, user could also pull using a digest

`quay.io/fosterseth/awx-ee@sha256:cbb1256806ea16762ea157bfe52578070b900b607780eca196f57838dd2d7eff`